### PR TITLE
implement RadioButtons

### LINKED ISSUE #27: Radio Buttons (Vertical and Horizontal)

### PR DESCRIPTION
**Pull request by barzamin**
Fixes #27.

### CODE REVIEW

File: iui/src/controls/entry.rs [hunk 1]
```diff
@@ -291,3 +292,46 @@ impl Checkbox {
         }
     }
 }
+
+define_control! {
```
Comment (NoraCodes) [inline iui/src/controls/entry.rs:301]: Could you add a doc comment here that gives a quick explanation of the control? For instance, numeric entry has "Numerical entry which allows users to select a value by picking a location along a line."

**Review by NoraCodes — COMMENTED** [review]
Do you think it would be worth the time to redesign `NumericEntry` and `TextEntry` into a generic `Entry` trait? Technically `RadioButtons` could be `NumericEntry`, though that's not really congruent with the semantic implications of that trait, but it's definitely `Entry<i32>`. Plus, users could then wrap `RadioButtons` and make `Entry<CustomEnum>`.

On the other hand it would take time and might be overcomplicated.

**Comment by barzamin** [comment]
I feel like if we make `(Numeric|Text)Entry` more generic, we should instead split by `DiscreteEntry` and `FreeformEntry`? like, a ComboBox could be considered `DiscreteEntry<String>`... ugh. not really sure what the right way forward is; the question is really who maintains responsibility for mapping indices in the combobox container to strings. and that's a function of what level of abstraction we want the API to operate at, eg "push onto a stack of items and remember what strings you pushed where" vs something else.

**Comment by NoraCodes** [comment]
I like the `DiscreteEntry`/`FreeformEntry` dichotomy a lot, especially since we could then provide trait-level `new` functions if they were generic over two types, user-facing and programmer-facing.

That does raise the question of the appropriate level of abstraction, though. Perhaps we should remove the traits altogether.

Either way, I don't think this is something we have to tackle in your PR.

**Review by NoraCodes — APPROVED** [review]

**Comment by barzamin** [comment]
Removed [WIP] from title; I don't really have anything left to change.

**Comment by barzamin** [comment]
oh, i should probably update the changelog...

File: iui/src/controls/entry.rs [hunk 2]
```diff
@@ -291,3 +292,47 @@ impl Checkbox {
...
+    pub fn append(&self, _ctx: &UI, name: &str) {
+        let c_string = CString::new(name.as_bytes().to_vec()).unwrap();
+        unsafe { ui_sys::uiRadioButtonsAppend(self.uiRadioButtons, c_string.as_ptr()); }
+    }
+
+    pub fn selected(&self, _ctx: &UI) -> i32 {
+        unsafe { ui_sys::uiRadioButtonsSelected(self.uiRadioButtons) }
+    }
+
+    pub fn set_selected(&self, _ctx: &UI, idx: i32) {
```
Comment (barzamin) [inline iui/src/controls/entry.rs]: @LeoTindall should this be `&self` or `&mut self`? this is inconsistent across the codebase (weirdly, for, eg, `ComboBox`, `append` takes `&self` but `set_selected` takes `&mut self`)

Comment (NoraCodes) [reply iui/src/controls/entry.rs]: This is one of the 0.4 blockers I have in my internal tracker. The whole API is going to be standardized to use the correct ones and I'm not sure which way to do that at the moment. So, this should probably be `&mut` but if it isn't I'll change it later if appropriate.

**Pull request merged**